models/Tetris.py

```diff
@@ -125,5 +148,266 @@ def rotate(self):
+
+
+    # Calculate best move (piece drop) for given state
+    def get_best_move(self):
+        best_score = 999999
+        best_move = None 
+        copied_figure = self.figure
+        # calculate best move for each rotation
+        for rotate_num in rotation_map.keys():
+            # for every column drop piece and calculate score
+            for col in range(self.width - 3): # TODO: FIX RANGE? (cuts off early)
+                # simulate new states
+                copied_state = copy.deepcopy(self)
+                copied_state.figure = copied_figure
+                copied_state.figure.x = col
+                for _ in range(rotate_num):
+                    copied_state.figure.rotate()
+                print("COL", col)
+                while not copied_state.intersects():
+                    copied_state.figure.y += 1
+                copied_state.figure.y -= 1
+                copied_state.simulate_freeze()
+                # print("X: ", copied_state.figure.x)
+                # print("Y: ", copied_state.figure.y)
+                # print("Rotation: ", copied_state.figure.rotation)
+                # print("STATE: ")
+                copied_state.get_string_field()
+
+                score = sum(copied_state.get_score())
+                # print("SCORE: ", score)
+                # keep track of best move by LOWEST score
+                if score < best_score:
+                    best_score = score
+                    best_move = (copied_state.figure.x, copied_state.figure.rotation)
+        # (self.figure.x, self.figure.rotation)
+        return best_move, best_score
+    
+
+    # we want the lowest score/prioritize the lowest score
+    def calculate_all_heuristics(self, figure, dx):
+        field = copy.deepcopy(self.field)
+        score = 0
+
+        copied_figure = Figure(figure.x + dx, figure.y, figure.type, figure.color,
+                                figure.rotation)
+        # drop figure all the way to bottom and calculate score
+        while not self.intersects_with_figure(copied_figure, 0, 0):
+            copied_figure.y += 1
+        copied_figure.y -= 1
+
+        for i1 in range(4):
+            for j2 in range(4):
+                if i1 * 4 + j2 in copied_figure.image():
+                    field[i1 + copied_figure.y][j2 + copied_figure.x] = copied_figure.color
+
+        holes = 0
+        for r in range(self.height):
+            for c in range(self.width):
+                if field[r][c] == 0 and r > 0 and field[r - 1][c] == 0:
+                    holes += 1
+
+        height = self.get_height(field)
+
+        lines = 0
+        for i in range(1, self.height):
+            zeros = 0
+            for j in range(self.width):
+                if field[i][j] == 0:
+                    zeros += 1
+            if zeros == 0:
+                lines += 1
+        score -= lines ** 20
+
+        return score + height + holes
+
+
+    # TODO: Fix??
+    def get_height(self, field):
```

Comment:
Is this right lol?

Reply:
I fixed it in my branch, didn't commit yet i think

Reply:
Oh my get_heights is right now though